### DISCUSSION TITLE
Help with MACR code formulation for scripting a tool

### DISCUSSION THREAD

**Discussion by IRCGraphic** [gopro/labs#776] [Q&A]
I've been tinkering with MACR and I would like to make a script to automatically generate the coefficients for a given lens distortion; I've seen that every coefficient (a,b,c..) responds to different distortion models:
![LENS PROBLEM](https://github.com/gopro/labs/assets/113941057/2ea4dc8f-6eca-45d9-b689-6cc87188f967)
Can anyone help me know what is the math that is used here so I can build a script? Of course I would share this tool for anyone who wants to use it, this will help make GoPro compatible with any sort of lens add-ons, thanks

**Comment by dnewman-gpsw** [comment]
This feature is not likely to be ported to future cameras, as it was from an incomplete hackathon project.  And the math is hard.

Lens distortions are radial r' = A.r + B.r^2 + C.r^3 etc.

MACR adds the new distortion in the internal lens model.

Making the math  r' = (A+a).r + (B+b).r^2 + (C+c).r^3 etc.

**Reply by IRCGraphic** [reply]
"And the math is hard." => On this subject there's a lot of documentation online, the problem is there's many different geometry methods, so which one is GoPro Labs implementing if I may ask?
Also working with MACR recipes, it seems to only work with stabilisation if the correction defaults to the perspective originally was selected on the GoPro? so "Wide" + Add-on => MACR correction => "Wide" field of view and perspective as result

**Reply by dnewman-gpsw** [reply]
This what GoPro uses r' = A.r + B.r^2 + C.r^3 ... etc.  radial all powers, not just odd.
I don't follow the reset of your reply.

**Reply by IRCGraphic** [reply]
Looks like standard models online show different math with each one of the coefficient types, as in the image A type is clearly very different form the B type, therefore I assume a complex math is being used beneath the a,b,c formula differently for each one of the coefficients?
Of course I tried before asking your formula without success, I'm no specialist in this stuff only a Gopro consumer trying to use Labs.

**Reply by dnewman-gpsw** [reply]
I've explained all I can.  You will not able to used online calibration tools, as this is a delta on top of the lens distortion models in camera.

**Reply by IRCGraphic** [reply]
Yeah I forgot, the script should also calibrate the default GoPo lens, delta you said?, and use it as a default to interpret the distortion added on top.
Like you said, perhaps is best to leave it as is, very interesting nonetheless, a Labs tool to calibrate add-on lenses is a realistic idea? I mean like a proper Labs calibration tool where once the lens profile/recipe is created one could chose the distortion/correction amount and zoom/field of view?

**Reply by dnewman-gpsw** [reply]
It is little outside of my area.  You need a solution that solves in the format r' = A.r + B.r^2 + C.r^3  , I've not seen one that does directly.  Most used the OpenCV model r' = A.r + C.r^3 + C.r^5 ...   But if you do get the the correct power function, the scale is likely the next issue. You could curve fit an OpenCV model to the required polynomial. You would solve for the stock lens, and the stock lens plus macro, subtract the two and send the delta to MACR.

**Reply by IRCGraphic** [reply]
I will look into that, it's definitely miles away from my area but just recently I found a way of creating MTRX recipes with Davinci so i thought perhaps I could push the luck with MACR.
I'd like to investigate further the MTRX benefits, perhaps I'll do a tutorial or something, made some progress with lowlight.

**Reply by IRCGraphic** [reply]
Only to share, a quick sketch of the MTRX with Davinci Resolve method, yellow street lights and night sky blue noise are very problematic in post so this MTRX code compresses all primaries except cyan, cyan seems to behave ok in lowlight, the corrective LUT is not meant for accuracy rather a creative method to keep chromatic noise down while mantaining rich pallette:
![mtrx_sample](https://github.com/gopro/labs/assets/113941057/48f59aec-03db-46ac-ab84-655958c6f772)
